TEMPLATE_ENGINE.md

```diff
@@ -0,0 +1,126 @@
+# Template Engine
```

Comment:
When I was revisiting some of the code — I found myself wishing a document like this existed. Additionally, I think that having to describe some of this stuff _in prose_ has helped me refactor / rename the actual code to be more self-documenting.

This does get into some nitty-gritty implementation details, but I think it might be worth keeping around for developer perusal.